fdo.go

```diff
@@ -0,0 +1,264 @@
+}
+
+//FdoLookupXdgDataDirs returns a string slice of all XDG_DATA_DIRS
+func FdoLookupXdgDataDirs() []string {
+	dataLocation := os.Getenv("XDG_DATA_DIRS")
+	locationLookup := strings.Split(dataLocation, ":")
+	if len(locationLookup) == 1 && locationLookup[0] == dataLocation {
+		var fallbackLocations []string
+		fallbackLocations = append(fallbackLocations, "/usr/local/share")
+		fallbackLocations = append(fallbackLocations, "/usr/share")
+		return fallbackLocations
+	}
+	return locationLookup
+}
+
+//FdoLookupApplication looks up an application by name and returns a FdoDesktop
+func FdoLookupApplication(appName string) *FdoDesktop {
+	locationLookup := FdoLookupXdgDataDirs()
+	for _, dataDir := range locationLookup {
+		testLocation := filepath.Join(dataDir, "applications", appName+".desktop")
+		if _, err := os.Stat(testLocation); os.IsNotExist(err) {
+			continue
+		} else {
+			return NewFdoDesktop(testLocation)
+		}
+	}
+	return nil
+}
+
+//FdoLookupApplicationWinInfo looks up an application based on window info
+func FdoLookupApplicationWinInfo(title string, classes []string, command string, iconName string) *FdoDesktop {
+	desktop := FdoLookupApplication(title)
+	if desktop != nil {
+		return desktop
+	}
+	for _, class := range classes {
+		desktop := FdoLookupApplication(class)
+		if desktop != nil {
+			return desktop
+		}
+	}
+	desktop = FdoLookupApplication(command)
+	if desktop != nil {
+		return desktop
+	}
+	desktop = FdoLookupApplication(iconName)
+	return desktop
+}
+
+func reverseWalkDirectoryMatch(directory string, joiner string, iconName string) string {
+	//If the requested icon wasn't found in the Example 32x32 or scalable dirs of the theme, try other sizes within theme
+	files, err := ioutil.ReadDir(directory)
+	if err != nil {
+		fyne.LogError("", err)
+		return ""
+	}
+	//Lets walk the files in reverse so bigger icons are selected first (Unless it is a 3 digit icon size like 128)
+	for i := len(files) - 1; i >= 0; i-- {
+		f := files[i]
+		if strings.HasPrefix(f.Name(), ".") == false {
+			if f.IsDir() {
+				matchDir := filepath.Join(directory, f.Name())
+				for _, extension := range iconExtensions {
+					//Example is /usr/share/icons/icon_theme/64x64/apps/xterm.png
+					var testIcon = ""
+					if joiner != "" {
+						testIcon = filepath.Join(matchDir, joiner, iconName+extension)
+					} else {
+						testIcon = filepath.Join(matchDir, iconName+extension)
+					}
+					if _, err := os.Stat(testIcon); os.IsNotExist(err) {
+						continue
+					} else {
+						return testIcon
+					}
+				}
+			}
+		}
+	}
+	return ""
+}
+
+func lookupIconPathInTheme(dir string, iconName string) string {
+	if _, err := os.Stat(dir); os.IsNotExist(err) {
+		return ""
+	}
+	for _, extension := range iconExtensions {
+		//Example is /usr/share/icons/icon_theme/32x32/apps/xterm.png
+		testIcon := filepath.Join(dir, string(fyconSize), "apps", iconName+extension)
+		if _, err := os.Stat(testIcon); os.IsNotExist(err) {
```

Comment:
This is lots of repeated code each indenting further. As immediately above, reverse the logic then you can return in the if block - otherwise move on to the next line...?